### PR TITLE
More sanitization of submitted archives

### PR DESCRIPTION
**Pull request by blerner**
Need to lock down file size, extracted file size, extracted file count, extracted file locations

### CODE REVIEW

**Comment by blerner** [comment]
(Note: tests are temporarily failing; I'll fix that later tonight.)

**Comment by blerner** [comment]
FWIW, I dont understand why this test is failing: when I run the exact same test, with the same random seed, it passes on my machine.

File: app/models/upload.rb [hunk 1]
```diff
@@ -70,28 +76,76 @@ def count_contents(upload, meta)
...
       end
+      raise Exception.new("Could not list the files in #{upload_path}") if !status.success?
+      return output.lines.count - offset
     rescue Exception => e
       puts e.message
       raise Exception.new("Could not read archive to count files")
     end
   end
 
+  def total_size_too_large?(upload, meta)
```
Comment (NatTuck) [inline app/models/upload.rb]: An alternate plan to consider is making a 20MB RAM disk and unzipping there. If it fails, it was too big. This has the advantages of being archive format independent and avoids the two step "check and then unzip" process which is prone to logic errors.

File: app/models/upload.rb [hunk 2]
```diff
@@ -100,19 +154,22 @@ def extract_contents!(mimetype)
     extracted_path.mkpath
     if mimetype == "application/zip" ||
        submission_path.to_s.ends_with?(".zip")
-      ZipRuby::Archive.open(submission_path.to_s) do |ar|
-        ar.each do |zf|
-          if zf.directory?
-            FileUtils.mkdir_p(extracted_path.join(zf.name))
+      output, err, status = Open3.capture3("unzip", "-n", # Never overwrite
```
Comment (NatTuck) [inline app/models/upload.rb]: It might be worth pulling all this archive logic out into an archive handling module in /lib

File: lib/sub_tarball.rb [hunk 3]
```diff
@@ -120,19 +103,44 @@ def self.untar_source(source, destination)
...
+            FileUtils.mkdir_p(File.dirname(dest))
+            File.open dest, "wb" do |f|
+              f.print entry.read
+            end
+            FileUtils.chmod entry.header.mode, dest, :verbose => false
+          elsif entry.header.typeflag == '2' #Symlink!
+            # skip for now, come back to it after all other files have been created
+          end
+        else
+          raise Exception.new("Could not create #{entry.full_name}: this path does not stay within the submission directory.")
```
Comment (NatTuck) [inline lib/sub_tarball.rb]: Stupid symlinks.

File: app/assets/javascripts/grades.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      e.preventDefault();
+      e.stopPropagation();
+      selectTreeviewFileByHref($(this).attr("href"));
+    });
+  }
+
+
+  run_on_page("grades/show", initFiles);
+  run_on_page("grades/edit", initFiles);
+  run_on_page("grades/show", initFiles);
```
Comment (blerner) [inline app/assets/javascripts/grades.js]: No, fixed.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -204,6 +194,41 @@ def fix_hrefs(node)
     fix_hrefs({nodes: @submission_files})
   end
 
+  def mime_type(full_path)
```
Comment (blerner) [inline app/controllers/application_controller.rb]: Maybe, though it's misleadingly named at the moment (it's not quite true MIME types).  I'll clean it up in a separate pass, probably

File: app/controllers/files_controller.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    if File.file?(Upload.base_upload_dir.join(params[:path]))
+      case File.extname(params[:path]).downcase
+      when ".jpg", ".jpeg", ".png", ".gif"
+        disp = "inline"
+      else
+        disp = "attachment"
+      end
+      mime = mime_type(params[:path])
+      
+      send_file Upload.base_upload_dir.join(params[:path]).to_s,
```
Comment (blerner) [inline app/controllers/files_controller.rb]: Say more?  I don't understand how to do that.

File: app/helpers/course_spreadsheet.rb [hunk 7]
```diff
@@ -229,12 +229,7 @@ def sanity_check
   end
 
   def sanitize(str)
-    str = str.strip
-    if str[0] == "=" || str[0] == "+" || str[0] == "-" || str[0] == "@"
-      str[1..str.length]
-    else
-      str
-    end
+    str.gsub(/^\s+[-+=@]+/, "").strip
```
Comment (blerner) [inline app/helpers/course_spreadsheet.rb]: Bug, fixed.

File: app/models/upload.rb (hunk 1 above)
Comment (blerner) [inline app/models/upload.rb]: Yes, that was my current thought.  It's not quite trivial to set up, and it doesn't solve the symlink issues, though, so I think there's still work to be done here.

File: app/models/upload.rb [hunk 8]
```diff
@@ -70,28 +76,76 @@ def count_contents(upload, meta)
...
+    if upload.is_a? ActionDispatch::Http::UploadedFile
+      upload_path = upload.path
+    elsif upload.is_a? String
+      upload_path = upload
+    else
+      upload_path = submission_path.to_s
+    end
+    begin
+      if meta[:mimetype] == "application/zip" || upload_path.ends_with?(".zip")
+        output, err, status = Open3.capture3("unzip", "-l", upload_path)
```
Comment (blerner) [inline app/models/upload.rb]: I don't know.  I've seen one misleading .zip file whose reported sizes are way too _large_; I don't know if it's possible for them to be misreported as too _small_.

File: app/models/upload.rb [hunk 9]
```diff
@@ -70,28 +76,76 @@ def count_contents(upload, meta)
...
+        return File.size(upload_path) > Upload.MAX_SIZE
+      elsif meta[:mimetype] == "application/x-compressed-tar" || upload_path.ends_with?(".tgz") ||
+            meta[:mimetype] == "application/gzip" || upload_path.ends_with?(".gz")
+        # try getting the size info directly from the gzipped file (but size is mod 2^32)
+        output, err, status = Open3.capture3("gunzip", "-l", upload_path)
+        raise Exception.new("could not list the files in #{upload_path}") if !status.success?
+        return true if output.lines.last.split.second.to_i > Upload.MAX_SIZE
+        # if the size is "small", it might've actually been greater than 2^32 and overflowed
+        # so try the slower, manual approach
+        Zlib::GzipReader.open(upload_path) do |zf|
```
Comment (blerner) [inline app/models/upload.rb]: Thanks.  It made me groan, having to write it....

File: app/models/upload.rb (hunk 2 above)
Comment (blerner) [inline app/models/upload.rb]: Yup.  I haven't figured out the clean interface that abstracts over Zlib::GzipReader and Zip::File and TarReader and whatnot, yet, though.  I want to; it's on my todo list.

File: lib/sub_tarball.rb (hunk 3 above)
Comment (blerner) [inline lib/sub_tarball.rb]: Yup.

File: app/controllers/files_controller.rb (hunk 6 above)
Comment (NatTuck) [inline app/controllers/files_controller.rb]: We could keep /public and then tell nginx that it should serve the files there as application/octet-stream, content-disposition: attach.

There are probably browser edge cases though, and I'm not sure what happens if you  to an attachment.

**Comment by NatTuck** [comment]
The test failure is showing up as 

"Team.count" didn't change by 0.
Expected: 11
Actual: 21

Looks like it's adding ten teams when you expected zero.

**Comment by blerner** [comment]
Yes, I saw the test error on travis.  But when I run the exact same test in the exact same way with the exact same test seed on _my_ machine, the test passes.

File: app/controllers/files_controller.rb (hunk 6 above)
Comment (blerner) [inline app/controllers/files_controller.rb]: I checked that, actually: `` is fine, and the image will show up.  But there's one case where it's less fine: If a student submits a large image that doesn't fit on the screen, then the file-viewing UI will show it compressed to fit, with a magnifying-glass icon to pop it out.  That is simply a `<a target="_blank" href="image path">`, and that will open a new tab showing the image.  If content-disposition=attachment, though, it'll open a new tab and instantly close it and offer a Download File dialog, which is not the ideal experience.

File: lib/archive_utils.rb [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+            FileUtils.rm_rf dest unless File.file? dest
+            FileUtils.mkdir_p(File.dirname(dest))
+            # skip creating the symlink for now
+          end
+        else
+          raise SafeExtractionError.new(file, entry.full_name)
+        end
+        dest = nil
+      end
+      # Now go through again, only for creating the symlinks
```
Comment (NatTuck) [inline lib/archive_utils.rb]: Would it be simpler and similarly useful to just skip symlinks?

I can't think of a case where I'd want symlinks in the unpacked directory for manual grading.

Comment (blerner) [inline lib/archive_utils.rb]: I'm not sure, honestly.  I'm including them (a) because I can imagine a "basic shell competency" course, where a plausible assignment is to demonstrate that you can construct a particular directory/file structure, and more importantly (b) because I tend to regret not being thorough and having "semantically-complete" implementations later on, when I wonder "why doesn't Thing X work?" and have to reconstruct the engineering reasons behind the implementation's limits.

I can clean up the efficiency of it, a bit, by setting a flag saying "yes, we've seen a symlink", and only traversing the archive again if there were any symlinks to be handled.
